libp2p-swarm/src/transport.rs

```diff
@@ -164,6 +184,48 @@ impl<A, B> Transport for OrTransport<A, B>
 	}
 }
 
+/// Implementation of `ConnetionUpgrade`. See `with_simple_protocol_upgrade`.
+#[derive(Debug)]
+pub struct SimpleProtocolUpgrade<F> {
+	name: Bytes,
+	// Note: we put the closure `F` in an `Arc` because Rust closures aren't automatically clonable
+	// yet.
+	upgrade: Arc<F>,
```

Comment:
I would agree with implementing the trait on a tuple, however there's the small problem that the closure has to be put in an `Arc` (because it has to be clonable, and closures aren't clonable in Rust yet).

I would suggest keeping `SimpleProtocolUpgrade`, but remove `with_simple_upgrade()` and let the user do `with_upgrade(SimpleProtocolUpgrade::new(...))` instead.

Reply:
Just make it `protocol_upgrade::Simple` so you don't have to repeat yourself so much (I've written similar things in other PRs).

Reply:
What about `SimpleProtocol`?